cdap-watchdog/src/test/java/co/cask/cdap/metrics/process/MetricsProcessorServiceTest.java

```diff
@@ -0,0 +1,300 @@
+    // Publish metrics to Kafka and record expected metrics before kafkaMetricsProcessorService starts
+    for (int i = 1; i < 10; i++) {
+      addKafkaMetrics(i, metricsContext, expected, preparer);
+    }
+    preparer.send();
+
+    // Sleep to make sure metrics get published
+    Thread.sleep(2000);
+
+    // Consume and process metrics after they are published to Kafka
+    KafkaMetricsProcessorService kafkaMetricsProcessorService =
+      new KafkaMetricsProcessorService(kafkaClient,
+                                       injector.getInstance(MetricDatasetFactory.class),
+                                       new MetricsMessageCallbackFactory(injector.getInstance(SchemaGenerator.class),
+                                                                         injector.getInstance(DatumReaderFactory.class),
+                                                                         metricStore, 4), topicPrefix, partitions);
+
+    kafkaMetricsProcessorService.setMetricsContext(context);
+    kafkaMetricsProcessorService.startAndWait();
+
+    // Intentionally set fetcher persist threshold to a small value, so that MessagingMetricsProcessorService
+    // internally can persist metrics when more messages are to be fetched
+    MessagingMetricsProcessorService messagingMetricsProcessorService =
+      new MessagingMetricsProcessorService(injector.getInstance(MetricDatasetFactory.class), topicPrefix,
+                                           partitions, messagingService, injector.getInstance(SchemaGenerator.class),
+                                           injector.getInstance(DatumReaderFactory.class),
+                                           metricStore,
+                                           1);
+
+
+
+    messagingMetricsProcessorService.setMetricsContext(context);
+    messagingMetricsProcessorService.startAndWait();
+
+    // Publish metrics and record expected metrics
+    for (int i = 10; i < 20; i++) {
+      publishMessagingMetrics(i, metricsContext, expected);
+    }
+
+    Thread.sleep(5000);
+    // Stop and restart messagingMetricsProcessorService
+    messagingMetricsProcessorService.stopAndWait();
+    // Intentionally set fetcher persist threshold to a large value, so that MessagingMetricsProcessorService
+    // internally only persists metrics during terminating.
+    messagingMetricsProcessorService =
+      new MessagingMetricsProcessorService(injector.getInstance(MetricDatasetFactory.class), topicPrefix,
+                                           partitions, messagingService, injector.getInstance(SchemaGenerator.class),
+                                           injector.getInstance(DatumReaderFactory.class),
+                                           metricStore,
+                                           100);
+    messagingMetricsProcessorService.startAndWait();
+
+    for (int i = 20; i < 30; i++) {
+      publishMessagingMetrics(i, metricsContext, expected);
+    }
+
+    Thread.sleep(5000);
```

Comment:
rather than unconditionally sleep for 5 seconds, you should use Tasks.waitFor() with a timeout of 5 seconds.